### PR TITLE
feat(cli): implement memory subcommand for conversation management

### LINKED ISSUE #754: [FEATURE] Implement neurolink memory CLI subcommand for conversation memory management
## Problem Statement

**Is your feature request related to a problem? Please describe.**

The documentation at `docs/features/conversation-history.md` describes CLI commands for managing conversation memory (list, export, delete sessions), but these commands do not actually exist in the CLI implementation. Users reading the documentation expect to use commands like `neurolink memory list` or `neurolink memory export --session-id <ID>`, but receive "command not found" errors.

This creates a documentation-implementation gap and prevents users from managing their conversation sessions via the CLI.

## Proposed Solution

**Describe the solution you'd like**

Implement a `memory` subcommand in the NeuroLink CLI with the following commands:

### Commands to Implement

| Command | Description |
|---------|-------------|
| `neurolink memory list` | List all conversation sessions with metadata |
| `neurolink memory export --session-id <ID>` | Export a specific session to JSON/CSV |
| `neurolink memory export-all` | Export all sessions to a directory |
| `neurolink memory delete --session-id <ID>` | Delete a specific session |
| `neurolink memory clear` | Clear all sessions (with confirmation) |

### API Design

```bash
# List all sessions for current user
neurolink memory list
# Output format:
# session-123 | "Machine Learning Discussion" | 15 messages | last active: 2024-01-01
# session-456 | "API Design Review"          |  8 messages | last active: 2024-01-02

# List with JSON output
neurolink memory list --format json

# Export single session to JSON
neurolink memory export --session-id session-123 --format json > history.json

# Export with full metadata
neurolink memory export --session-id session-123 --include-metadata

# Export to CSV format
neurolink memory export --session-id session-123 --format csv > history.csv

# Export all sessions to directory
neurolink memory export-all --output ./exports/

# Delete specific session
neurolink memory delete --session-id session-123

# Delete with force (skip confirmation)
neurolink memory delete --session-id session-123 --force

# Clear all sessions
neurolink memory clear --confirm

# Get memory stats
neurolink memory stats
```

### Expected Inputs and Outputs

**Input Options:**
- `--session-id <string>`: Session identifier (required for export/delete)
- `--format <json|csv>`: Output format (default: json)
- `--include-metadata`: Include session metadata in export
- `--output <path>`: Output directory for export-all
- `--confirm`: Skip confirmation prompt for destructive operations
- `--force`: Force operation without prompts
- `--user-id <string>`: User ID for multi-user scenarios (optional)

**Output:**
- Formatted table output for list command
- JSON/CSV file output for export commands
- Confirmation messages for delete/clear operations

## Example Usage

**Describe how you would use this feature**

```typescript
// This is CLI usage, not SDK - showing the file structure needed

// src/cli/commands/memory.ts - New file to create
import { CommandModule } from "yargs";
import { createMemoryCommands } from "../factories/memoryCommandFactory.js";

export const memoryCommand: CommandModule = {
  command: "memory <subcommand>",
  describe: "Manage conversation memory and session history",
  builder: (yargs) => createMemoryCommands(yargs),
  handler: () => {},
};
```

```bash
# CLI usage examples

# Developer workflow: Debug a problematic conversation
neurolink memory list --format json | jq '.[] | select(.title | contains("error"))'
neurolink memory export --session-id session-problematic --include-metadata > debug.json

# Operations workflow: Clean up old sessions
neurolink memory list
neurolink memory delete --session-id old-session-1 --force
neurolink memory delete --session-id old-session-2 --force

# Analytics workflow: Export all for analysis
neurolink memory export-all --output ./weekly-export/ --format csv
```

## Alternatives Considered

**Describe alternatives you've considered**

1. **Direct Redis CLI access**: Users could use `redis-cli` to query session data directly
   - Why insufficient: Requires Redis knowledge, doesn't handle serialization/deserialization, no user-friendly formatting

2. **SDK-only API**: Keep memory management as SDK-only feature
   - Why insufficient: CLI users expect parity with documented features, CLI is often used for debugging/operations

3. **Loop mode commands**: Add memory commands as loop-mode internal commands (`/memory list`)
   - Why insufficient: Users expect these to work as standalone CLI commands matching documentation

## Use Case

**What is your use case?**

- **Industry**: Enterprise AI development (all industries using NeuroLink)
- **Application type**: Developer tooling, DevOps, debugging, compliance auditing
- **Scale**: Production - teams managing multiple conversation sessions

**Specific scenarios:**
1. **Debugging**: Developer needs to export a conversation to understand why AI gave unexpected response
2. **Compliance**: Team needs to export all conversations for audit trail
3. **Operations**: Ops engineer needs to clean up stale sessions to manage Redis storage
4. **Analytics**: Data team needs to export conversations for quality analysis

## Impact

**Who would benefit from this feature?**

- [x] Individual developers
- [x] Small teams
- [x] Enterprise users
- [x] All users

### Current State

The SDK already has the underlying implementation in `RedisConversationMemoryManager`:

```typescript
// Existing methods in src/lib/core/redisConversationMemoryManager.ts
getUserSessions(userId: string): Promise<string[]>
getUserSessionMetadata(userId: string, sessionId: string): Promise<SessionMetadata | null>
getUserSessionHistory(userId: string, sessionId: string): Promise<ChatMessage[] | null>
getUserSessionObject(userId: string, sessionId: string): Promise<RedisConversationObject | null>
getUserAllSessionsHistory(userId: string): Promise<SessionMetadata[]>
clearSession(sessionId: string, userId?: string): Promise<boolean>
clearAllSessions(): Promise<void>
getStats(): Promise<ConversationMemoryStats>
```

### Documentation Reference

The following documentation already describes these CLI commands:
- `docs/features/conversation-history.md` - Lines 76-80, 266-269

### Related Files

| File | Purpose |
|------|---------|
| `src/lib/core/redisConversationMemoryManager.ts` | Existing Redis memory manager with required methods |
| `src/cli/commands/` | Directory for CLI command implementations |
| `src/cli/factories/commandFactory.ts` | Pattern to follow for command creation |
| `docs/features/conversation-history.md` | Documentation to align with |

## Implementation Considerations

**Technical details**

- **Affected components**: CLI only (SDK methods already exist)
- **Breaking changes**: No - this adds new functionality
- **Provider compatibility**: N/A - this is storage layer (Redis)
- **Performance implications**: 
  - `list` command should use metadata-only fetch (already optimized)
  - `export-all` should handle pagination for large session counts
  - `clear` should use batch deletion (already implemented)

### Suggested Implementation Steps

1. Create `src/cli/commands/memory.ts` with subcommand definitions
2. Create `src/cli/factories/memoryCommandFactory.ts` following existing patterns
3. Register memory command in `src/cli/index.ts`
4. Add proper error handling for Redis connection issues
5. Implement table formatting for list output
6. Add confirmation prompts for destructive operations
7. Add tests in `test/cli/memory.test.ts`
8. Update CLI documentation in `docs/cli/commands.md`

### Example Output Formats

**List command (table format):**
```
Session ID    | Title                        | Messages | Last Active
--------------|------------------------------|----------|---------------------
session-123   | Machine Learning Discussion  | 15       | 2024-01-01 10:30:00
session-456   | API Design Review            | 8        | 2024-01-02 14:15:00
```

**Export command (JSON format):**
```json
{
  "sessionId": "session-123",
  "title": "Machine Learning Discussion",
  "userId": "user-001",
  "createdAt": "2024-01-01T10:00:00Z",
  "updatedAt": "2024-01-01T10:30:00Z",
  "messages": [
    {
      "id": "msg-001",
      "role": "user",
      "content": "What is machine learning?",
      "timestamp": "2024-01-01T10:00:00Z"
    },
    {
      "id": "msg-002", 
      "role": "assistant",
      "content": "Machine learning is...",
      "timestamp": "2024-01-01T10:00:05Z"
    }
  ]
}
```

## Checklist

- [x] I have searched existing issues and feature requests to ensure this is not a duplicate
- [x] I have clearly described the problem and proposed solution
- [x] I have provided example usage code
- [x] I have considered alternatives and explained why they don't work
- [x] I have described the real-world use case

### PR DESCRIPTION
**Pull request by Harshita-Rupani29**
# Pull Request

## Description

**What does this PR do?**

Implements the `neurolink memory` CLI subcommand for managing conversation memory sessions. This feature closes the documentation-implementation gap where CLI commands were documented in `docs/features/conversation-history.md` but not actually implemented.

## Related Issues

**Does this PR close any issues?**

Closes #754

## Type of Change

Please select the type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [x] Test coverage improvement
- [ ] Build/CI configuration
- [ ] Other (please describe):

## Motivation and Context

**Why is this change needed? What problem does it solve?**

The documentation at `docs/features/conversation-history.md` described CLI commands for managing conversation memory (list, export, delete sessions), but these commands did not exist. Users reading the documentation expected to use commands like `neurolink memory list` or `neurolink memory export --session-id <ID>`, but received "command not found" errors.

This creates a documentation-implementation gap and prevents users from managing their conversation sessions via the CLI.

**Use cases:**
- **Debugging**: Developer needs to export a conversation to understand why AI gave unexpected response
- **Compliance**: Team needs to export all conversations for audit trail
- **Operations**: Ops engineer needs to clean up stale sessions to manage Redis storage
- **Analytics**: Data team needs to export conversations for quality analysis

## Changes Made

**What specific changes were made?**

### New Types
- Added `SessionListItem` type to `src/lib/types/conversation.ts` - Session metadata for listing
- Added `SessionExport` type to `src/lib/types/conversation.ts` - Complete session export format

### Interface Updates
- Added `listSessions(userId?: string): Promise<SessionListItem[]>` to `IConversationMemoryManager` interface

### Memory Manager Implementations
- Implemented `listSessions()` in `ConversationMemoryManager` (in-memory storage)
- Implemented `listSessions()` in `RedisConversationMemoryManager` (Redis storage)
- Added `formatTimeAgo()` helper method to both managers

### SDK Methods
- Added `listSessions(userId?)` to NeuroLink class
- Added `exportSession(sessionId, options)` to NeuroLink class
- Added `exportAllSessions(userId?, options)` to NeuroLink class

### CLI Commands
- Added `memory list` - List all conversation sessions with metadata
- Added `memory export --session-id <ID>` - Export single session to JSON
- Added `memory export-all --output <dir>` - Export all sessions to directory
- Added `memory delete --session-id <ID>` - Delete a specific session (with confirmation)
- Updated `memory clear` - Added `--confirm` flag requirement for safety
- Existing `memory stats` and `memory history` - Already implemented

### Documentation
- Updated `docs/features/conversation-history.md` - Removed "Planned Feature" notices, documented all CLI commands

### Tests
- Created `test/cli/memory-commands.test.ts` with 45 comprehensive test cases

## Breaking Changes

**Does this PR introduce breaking changes?**

- [x] No breaking changes
- [ ] Yes, breaking changes (describe below)

## Testing

**How has this been tested?**

- [x] Unit tests added/updated
- [x] Integration tests added/updated
- [ ] E2E tests pass
- [x] Manual testing completed
- [ ] Tested with multiple providers: N/A (storage layer, not provider-specific)
- [x] Tested on multiple platforms: macOS

### Test Coverage

- [x] All new code is covered by tests
- [x] Existing tests pass
- [x] Coverage percentage maintained or improved

**Test Results:** 45 tests passing

| Category | Tests | Coverage |
|----------|-------|----------|
| `memory list` | 4 | List all, empty, filter by user, memory not enabled |
| `memory export` | 4 | Export session, include metadata, non-existent, empty ID |
| `memory export-all` | 4 | Export all, filter by user, empty result, metadata |
| `memory delete` | 3 | Delete, non-existent, memory not enabled |
| `memory clear` | 2 | Clear all, memory not enabled |
| `memory stats` | 3 | Return stats, zero stats, memory not enabled |
| `memory history` | 3 | Return history, empty, memory not enabled |
| CLI structure | 2 | Command structure validation |
| Edge cases | 7 | Long IDs, special chars, Unicode, concurrent ops, timeouts |
| Output formats | 2 | Structure validation |
| ConversationMemoryManager | 6 | listSessions unit tests |

### Manual Testing Steps

1. Build CLI: `pnpm run build:cli`
2. List sessions: `node dist/cli/index.js memory list`
3. Check help: `node dist/cli/index.js memory --help`
4. Verify all subcommands are registered
5. Run tests: `pnpm test test/cli/memory-commands.test.ts`

## Code Quality

**Have you followed code quality standards?**

- [x] Code follows the project's style guidelines (ESLint passes)
- [x] Code is properly formatted (Prettier applied)
- [x] Self-review of code completed
- [x] No console.log statements (using logger instead)
- [x] No hardcoded API keys or secrets
- [x] TypeScript strict mode compliance
- [x] Proper error handling implemented
- [x] TODO/FIXME comments reference issues

## Documentation

**Have you updated documentation?**

- [x] JSDoc comments added/updated for public APIs
- [ ] README.md updated (if needed)
- [x] Documentation in /docs updated (if needed)
- [x] Code examples added/updated (if needed)
- [ ] CHANGELOG.md updated (if applicable)
- [ ] Migration guide provided (if breaking changes)

## Commit Message Format

**Does your commit follow semantic commit conventions?**

- [x] Commit message follows format: `type(scope): description`
- [x] Valid type used: feat, fix, docs, style, refactor, test, chore, build, ci, perf, revert
- [x] Scope specified (e.g., providers, cli, docs, middleware)

Commit: `feat(cli): implement memory subcommand for conversation management (#754)`

## Dependencies

**Does this PR add, update, or remove dependencies?**

- [x] No dependency changes
- [ ] Dependencies added (list below)
- [ ] Dependencies updated (list below)
- [ ] Dependencies removed (list below)

## Performance Impact

**Does this change affect performance?**

- [x] No performance impact
- [ ] Performance improved (provide metrics)
- [ ] Performance degraded (justify why acceptable)

## Security Considerations

**Are there any security implications?**

- [x] No security implications
- [ ] Security review needed
- [ ] Security vulnerability fixed

The `memory delete` command includes a confirmation prompt by default to prevent accidental deletion. Use `--force` to skip confirmation.

## Deployment Notes

**Special deployment instructions?**

- [x] No special deployment steps
- [ ] Requires environment variable changes (list below)
- [ ] Requires database migration
- [ ] Requires Redis schema update
- [ ] Other (describe below)

## Reviewer Checklist

For reviewers:

- [ ] Code follows project style and conventions
- [ ] Changes are well-documented
- [ ] Tests provide adequate coverage
- [ ] No obvious performance issues
- [ ] No security vulnerabilities introduced
- [ ] Breaking changes are properly documented
- [ ] Documentation is clear and accurate

## Additional Notes

**Files Changed:**

| File | Changes |
|------|---------|
| `src/lib/types/conversation.ts` | Added `SessionListItem`, `SessionExport` types |
| `src/lib/types/conversationMemoryInterface.ts` | Added `listSessions` method to interface |
| `src/lib/core/conversationMemoryManager.ts` | Implemented `listSessions`, `formatTimeAgo` |
| `src/lib/core/redisConversationMemoryManager.ts` | Implemented `listSessions`, `formatTimeAgo` |
| `src/lib/neurolink.ts` | Added `listSessions`, `exportSession`, `exportAllSessions` |
| `src/cli/factories/commandFactory.ts` | Added memory list/export/export-all/delete commands |
| `docs/features/conversation-history.md` | Updated CLI documentation |
| `test/cli/memory-commands.test.ts` | Created with 45 test cases |
| `memory-bank/` | Updated progress tracking |

---

## Pre-submission Checklist

Before submitting, ensure you have:

- [x] Read and followed the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] Verified all automated pre-commit checks pass
- [x] Tested changes locally with `pnpm test`
- [x] Built the project successfully with `pnpm build`
- [x] Run `pnpm run validate:all` and all checks pass
- [x] Reviewed your own code for obvious issues
- [x] Ensured commit messages follow semantic format
- [x] Updated relevant documentation
- [x] Added tests for new functionality
- [ ] Checked that CI/CD pipeline passes (after creating PR)

Thank you for contributing to NeuroLink!

## Release Notes

* **New Features**
  * Added memory CLI subcommands for listing, exporting, and managing conversation sessions
  * SDK methods now support session operations: list, export single/all sessions with optional metadata
  * Redis-backed storage enables advanced memory and session export capabilities
  * New CLI commands: `memory list`, `memory export`, `memory export-all`, `memory delete`, `memory clear`, `memory stats`, `memory history`

* **Documentation**
  * Updated guides with Redis configuration requirements and memory management examples
  * Added troubleshooting resources for session export workflows

### CODE REVIEW

File: src/cli/factories/commandFactory.ts [hunk 1]
```diff
@@ -1278,14 +1411,25 @@ export class CLICommandFactory {
...
                   "$0 memory clear session-123",
                   "Clear specific session",
                 ),
             async (argv) =>
               await CLICommandFactory.executeMemoryClear(
-                argv as BaseCommandArgs & { sessionId?: string },
+                argv as BaseCommandArgs & {
+                  sessionId?: string;
+                  confirm?: boolean;
+                },
```
[the remark this answers is not in the document]
Comment (Harshita-Rupani29) [reply src/cli/factories/commandFactory.ts:2343]: Done.

File: src/cli/factories/commandFactory.ts [hunk 2]
```diff
@@ -3145,6 +3289,392 @@ export class CLICommandFactory {
...
+          ),
+          resolve,
+        );
+      });
+      rl.close();
+
+      if (answer.toLowerCase() !== "y" && answer.toLowerCase() !== "yes") {
+        logger.always(chalk.gray("Deletion cancelled."));
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (Harshita-Rupani29) [reply src/cli/factories/commandFactory.ts]: Done.

File: src/lib/core/redisConversationMemoryManager.ts [hunk 3]
```diff
@@ -1315,6 +1316,65 @@ User message: "${userMessage}"`;
...
+          userId,
+          messageCount: 0, // Will need to fetch separately for accurate count
+          lastActive,
+        };
+      })
+      .sort(
+        (a, b) =>
+          new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime(),
+      );
+  }
```
[the remark this answers is not in the document]
Comment (Harshita-Rupani29) [reply src/lib/core/redisConversationMemoryManager.ts:1875]: Done.

File: src/lib/neurolink.ts [hunk 4]
```diff
@@ -6764,6 +6766,170 @@ Current user's request: ${currentInput}`;
...
+    }
+
+    // Check if listSessions is available on the memory manager
+    if (!this.conversationMemory.listSessions) {
+      logger.warn("listSessions not available on current memory manager");
+      return [];
+    }
+
+    try {
+      const sessions = await this.conversationMemory.listSessions(userId);
```
[the remark this answers is not in the document]
Comment (Harshita-Rupani29) [reply src/lib/neurolink.ts]: Done.

File: test/cli/memory-commands.test.ts [hunk 5]
```diff
@@ -0,0 +1,1046 @@
...
+        }),
+      );
+
+      const sdk = mockGetOrCreateNeuroLink();
+
+      await expect(sdk.clearAllConversations()).rejects.toThrow(
+        "Conversation memory is not enabled",
+      );
+    });
+  });
```
[the remark this answers is not in the document]
Comment (Harshita-Rupani29) [reply test/cli/memory-commands.test.ts]: Done.

File: test/cli/memory-commands.test.ts [hunk 6]
```diff
@@ -0,0 +1,1046 @@
...
+            .fn()
+            .mockRejectedValue(new Error("Redis connection refused")),
+        }),
+      );
+
+      const sdk = mockGetOrCreateNeuroLink();
+
+      await expect(sdk.listSessions()).rejects.toThrow(
+        "Redis connection refused",
+      );
```
[the remark this answers is not in the document]
Comment (Harshita-Rupani29) [reply test/cli/memory-commands.test.ts]: Done.

File: test/cli/memory-commands.test.ts
Comment (murdore) [inline test/cli/memory-commands.test.ts:1]: We need to move to continuous test suit only.

**Review by murdore — CHANGES_REQUESTED** [review]

File: test/cli/memory-commands.test.ts
Comment (Harshita-Rupani29) [reply test/cli/memory-commands.test.ts:1]: Done.

**Comment by murdore** [comment]
Thanks for this — it closes a real gap (the CLI docs still describe memory list/export/export-all/delete as "planned", and none of that exists on release yet), and CI is green. A maintainer requested moving tests into the continuous suite, which you've done. Before we can merge, could you: (1) rebase onto current release to resolve the merge conflicts, and (2) address the remaining CodeRabbit findings that are still open in the diff — in particular the path-traversal risk in executeMemoryExportAll (writing ${sessionExport.sessionId}.json directly into the output dir without sanitizing/validating it stays inside that dir) and the hardcoded format ternary that ignores --format csv? Happy to re-review once those are in. If we don't hear back in a couple of weeks we may close this as stale, but it's a welcome contribution and we'd like to land it.

**Comment by murdore** [comment]
🎉 Merged — thank you, @Harshita-Rupani29! This closes #754 and finally makes the documented `neurolink memory` commands real. I really appreciated how thoroughly you worked through the review — the path-traversal containment on `export-all`, `--format csv` actually flowing through, the `--confirm`/`--force` safety gates on the destructive paths, and moving the tests into `continuous-test-suite-memory.ts` to match the repo convention. That kind of iteration makes a PR a pleasure to merge.

Since you're clearly comfortable with NeuroLink's conversation-memory layer now — you could put it to work in a project of your own. A natural fit: a **personal CLI assistant or a Slack/Discord bot that remembers context across sessions**. Point `REDIS_URL` at a Redis instance and NeuroLink auto-persists conversations and auto-compacts them via the same summarization engine you just exposed, so a long-running assistant stays inside the model's context window on its own. Pass a stable `sessionId` to `neurolink.generate({ input, provider, context: { sessionId } })` and you've got a stateful, multi-provider agent in a few lines — swap `--provider anthropic` for `google-ai`/`openai`/etc. without touching your app code.

Would genuinely love to see what you build with it. Thanks again for the contribution! 🙌
